components/ws.js

```diff
@@ -0,0 +1,71 @@
+import { createContext, useContext, useEffect, useState, useCallback } from 'react'
+
+export const WebSocketContext = createContext(null)
+
+const WEBSOCKET_TIMEOUT = 30000
+
+export function WebSocketProvider ({ children }) {
+  const [ws, setWs] = useState(null)
+
+  useEffect(() => {
+    if (ws) return
+
+    // the endpoint needs to be hit at least once using HTTP
+    // to initialize the websocket server
+    fetch('http://localhost:3000/api/ws')
+
+    const client = new WebSocket('ws://localhost:3000/api/ws')
```

Comment:
Race condition here regarding the first fetch from the first user after a deploy.

There is probably a better way to initialize the websocket server instead of every user first fetching the endpoint.